### PR TITLE
sof: drc: Fixed-point component implementation

### PR DESCRIPTION
**Pull request by johnylin76**

### CODE REVIEW

File: src/audio/drc/drc_generic.c [hunk 1]
```diff
@@ -8,41 +8,46 @@
 #include <sof/audio/drc/drc.h>
 #include <sof/audio/drc/drc_math.h>
 #include <sof/audio/format.h>
+#include <sof/math/decibels.h>
 #include <sof/math/numbers.h>
 #include <stdint.h>
 
+#define ONE_Q20        Q_CONVERT_FLOAT(1.0f, 20)                /* Q12.20 */
```
Comment (cujomalainey) [inline src/audio/drc/drc_generic.c:15]: I would prefer we don't do this with the defines, I don't think we need to name these values. Keep them inline please (unless anyone objects)

**Review by cujomalainey — COMMENTED** [review]
LGTM but really should have @singalsu sign off as most of this fixed point macro math is beyond my knowledge.

**Review by lgirdwood — COMMENTED** [review]
@johnylin76 can you add some more detail in the commit message but otherwise LGTM.
@singalsu can you take a look here and provide any comments or approval.

**Comment by johnylin76** [comment]
> @johnylin76 can you add some more detail in the commit message but otherwise LGTM.

Added, thanks.

File: src/audio/drc/drc_generic.c [hunk 2]
```diff
@@ -450,23 +490,227 @@ static void drc_s16_default(const struct comp_dev *dev,
...
+		 * in multi-band DRC kernel case. */
+		for (ch = 0; ch < nch; ++ch) {
+			pd_write_index = state->pre_delay_write_index;
+			pd_read_index = state->pre_delay_read_index;
+			pd_write = (int32_t *)state->pre_delay_buffers[ch] + pd_write_index;
+			pd_read = (int32_t *)state->pre_delay_buffers[ch] + pd_read_index;
+			idx = ch;
+			for (i = 0; i < frames; ++i) {
+				x = audio_stream_read_frag_s32(source, idx);
+				y = audio_stream_read_frag_s32(sink, idx);
```
Comment (singalsu) [inline src/audio/drc/drc_generic.c:636]: Should this be audio_stream_write_frag_s32() and the same for previous function for s24 format?

**Review by singalsu — CHANGES_REQUESTED** [review]
Good work with fixed conversion! Please check few issues those I commented.

**Comment by cujomalainey** [comment]
Also see checkpatch errors

File: src/audio/drc/drc_math_generic.c [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+	/* log(x) = log(10) * log10(x) */
+	const int32_t LOG10 = Q_CONVERT_FLOAT(2.3025850929940457f, 29);
+	int32_t log10_x = Q_CONVERT_FLOAT(warp_log10f(x), 26);
+	return Q_CONVERT_QTOF(Q_MULTSR_32X32((int64_t)LOG10, log10_x, 29, 26, 26), 26);
+#else
+	return logf(x);
+#endif
+}
+
+inline float warp_sinf(float x)
```
[the remark this answers is not in the document]
Comment (johnylin76) [reply src/audio/drc/drc_math_generic.c]: Thanks. Used sin_fixed instead.

@singalsu Do you consider all other fixed implementation (e.g. asin, lin2db, log...) which are not existing in math library should be moved there?

**Comment by cujomalainey** [comment]
Still a bunch of checkpatch errors https://sof-ci.01.org/sofpr/PR3549/build7565/checkpatch/

**Comment by johnylin76** [comment]
> Still a bunch of checkpatch errors https://sof-ci.01.org/sofpr/PR3549/build7565/checkpatch/

I cleaned up all warnings except the "ABI update" one (don't know what to do for this).
Is it required to clean up "checks" as well?

File: src/audio/drc/drc_math_generic.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (singalsu) [reply src/audio/drc/drc_math_generic.c]: @johnylin76 Yes please, I think those functions are generic and very useful.

**Comment by cujomalainey** [comment]
The ending in ( line issue should probably be fixed. CamelCase, - macro reuse and - spacing look like they should be ignored.

**Comment by johnylin76** [comment]
> The ending in ( line issue should probably be fixed. CamelCase, - macro reuse and - spacing look like they should be ignored.

Fixed checks for "ending in ( line".

**Review by lgirdwood — APPROVED** [review]
@singalsu @cujomalainey please merge when you are good with the PR. It LGTM.

**Comment by johnylin76** [comment]
I rebased the PR to upstream due to https://github.com/thesofproject/sof/pull/3581 update.

**Comment by johnylin76** [comment]
> @singalsu @cujomalainey please merge when you are good with the PR. It LGTM.

@singalsu @cujomalainey friendly ping :)

**Comment by cujomalainey** [comment]
Waiting for @singalsu as he is the fixed point master :) Ill check tomorrow and sign off then

**Review by singalsu — APPROVED** [review]
Looks good, thanks @johnylin76 !

**Pull request merged**